### PR TITLE
planner/cascades: add transformation rule TransformAggregateCaseToSelection

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve? 
This PR adds Transformation rule TransformAggregateCaseToSelection for #14042.

### What is changed and how it works?
add a transformation rule TransformAggregateCaseToSelection.This rule tries to convert Agg(case when) to Agg->Selection.

### Check List

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

**Comment by SeaRise** [comment]
> A big problem still need to solve. If all of the `case when` return null value, the aggFunc should return its default value. But if we convert it to selection, we will lose some row. For example:
> 
> ```
> mysql> select * from tbl;
> +------+------+
> | a    | b    |
> +------+------+
> |    0 |    0 |
> |    0 |    1 |
> |    1 |    0 |
> +------+------+
> 3 rows in set (0.00 sec)
> mysql> select sum(case when b > 0 then b end) from tbl group by a;
> +---------------------------------+
> | sum(case when b > 0 then b end) |
> +---------------------------------+
> |                            NULL |
> |                               1 |
> +---------------------------------+
> 2 rows in set (0.00 sec)
> 
> mysql> select sum(b) from tbl where b > 0 group by a;
> +--------+
> | sum(b) |
> +--------+
> |      1 |
> +--------+
> 1 row in set (0.01 sec)
> ```

for `Agg(case when)`, we can add a restriction that `len(LogicalAggregation.GroupByItems) = 0`
```
mysql> select * from tbl;
+---+---+---+
| a | b | c |
+---+---+---+
| 0 | 0 | 0 |
| 0 | 1 | 0 |
| 1 | 0 | 0 |
+---+---+---+
3 rows in set (0.00 sec)

mysql> select sum(case when b > 0 then b end) from tbl;
+---------------------------------+
| sum(case when b > 0 then b end) |
+---------------------------------+
|                               1 |
+---------------------------------+
1 row in set (0.00 sec)

mysql> select sum(b) from tbl where b > 0;
+--------+
| sum(b) |
+--------+
|      1 |
+--------+
1 row in set (0.00 sec)
```

**Comment by SeaRise** [comment]
and we can add a new  rule
transformation `selection->Agg(case when)` to `selection->Agg->selection`.
and `len(agg.GroupByItems) != 0`.
rule can work only if `selection` make sure that `aggFunc` can not be Null or Zero.

such as
`select k from (select sum(case when b > 0 then b end as k) from tbl group by a) where k > 0`
can be convert to 
`select k from (select sum(b) as k from tbl where b > 0 group by a) where k > 0`

**Comment by francis0407** [comment]
Yeah, I think your idea works.
As for this PR, I think we can finish the `len(GroupByItems) == 0` part.

**Comment by SeaRise** [comment]
I have add some intergration test in `integration_test.go`.
`TransformAggregateCaseToSelection ` seem to work well in `TiKV`, 
and I add `EngineTiDBOrTiKV = EngineTypeSet(EngineTiDB | EngineTiKV)`.
Is it right?

**Comment by francis0407** [comment]
> I have add some intergration test in `integration_test.go`.
> `TransformAggregateCaseToSelection ` seem to work well in `TiKV`,
> and I add `EngineTiDBOrTiKV = EngineTypeSet(EngineTiDB | EngineTiKV)`.
> Is it right?

@SeaRise I prefer only matching the `CompleteMode` Aggregation, which only exists in EngineTiDB. Otherwise it may introduce too much complexity.

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -1781,3 +1783,123 @@ func isInnerJoinKeysContainUniqueKey(innerGroup *memo.Group, joinKeys *expressio
...
+		memo.OperandAggregation,
+		memo.EngineTiDBOnly,
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *TransformAggregateCaseToSelection) Match(expr *memo.ExprIter) bool {
+	agg := expr.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	return len(agg.GetGroupByUsedCols()) == 0 && len(agg.AggFuncs) == 1 && len(agg.AggFuncs[0].Args) == 1 && r.isTwoOrThreeArgCase(agg.AggFuncs[0].Args[0])
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: I think use `len(agg.GroupByItems) == 0` is ok? Do we have other cases?

Comment (francis0407) [inline planner/cascades/transformation_rules.go]: BTW, we'd better add the check `agg.IsCompleteModeAgg()` here?

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: group by constant? 
such as  select xxx from t group by 'a'

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: In `select xxx from t group by concat(a,c)`, `len(agg.GroupByItems) == 0` is true.

Comment (francis0407) [reply planner/cascades/transformation_rules.go]: But if we group by a constant, we cannot use this optimization?

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: `group by constant, no group by, group by func(constant...)` all is ok

Comment (francis0407) [reply planner/cascades/transformation_rules.go]: See the example:
```
mysql> select * from t;
+------+------+
| a    | b    |
+------+------+
|    1 |    1 |
+------+------+
1 row in set (0.00 sec)

mysql> select sum(b) from t where a > 1 group by 'p';
Empty set (0.00 sec)

mysql> select sum(case when a > 1 then b else 0 end) from t group by 'p';
+----------------------------------------+
| sum(case when a > 1 then b else 0 end) |
+----------------------------------------+
|                                      0 |
+----------------------------------------+
1 row in set (0.00 sec)
```

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: ok....

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: yes

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: yes, `len(agg.GroupByItems) == 0` is right

**Pull request merged**